### PR TITLE
cmd/action: add migrate diff command

### PR DESCRIPTION
**Pull request by masseelch**

### CODE REVIEW

File: cmd/action/migrate.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
+package action
```
Comment (a8m) [inline cmd/action/migrate.go:1]: Add license header.

File: cmd/action/migrate.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+		DevURL  string
+		ToURL   string
+		Schemas []string
+		Force   bool
+	}
+	// MigrateCmd represents the migrate command. It wraps several other sub-commands.
+	MigrateCmd = &cobra.Command{
+		Use:   "migrate",
+		Short: "Manage versioned migration files",
+		Long:  "`atlas migrate`" + ` wraps several sub-commands for migration management.`,
```
Comment (a8m) [inline cmd/action/migrate.go:37]:
```suggestion
		Long:  "'atlas migrate' wraps several sub-commands for migration management.`,
```

I changed it in apply, and I want to change in other places as well.

File: cmd/action/migrate.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+		PersistentPreRun: func(*cobra.Command, []string) {
+			// Migrate commands will not run on a broken migration directory, unless the force flag is given.
+			if !MigrateFlags.Force {
```
Comment (a8m) [inline cmd/action/migrate.go:40]:
```suggestion
			// Migrate commands will not run on a broken
			// migration directory, unless the force flag is given.
			if !MigrateFlags.Force {
```

File: cmd/action/migrate.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+				cobra.CheckErr(err)
+				cobra.CheckErr(migrate.Validate(dir)) // TODO(masseelch): tell the user what's wrong
+			}
+		},
+	}
+	// MigrateDiffCmd represents the migrate diff command.
+	MigrateDiffCmd = &cobra.Command{
+		Use:   "diff",
+		Short: "Compute the diff between the migration directory and a connected database and create a new migration file.",
+		Long: "`atlas migrate diff`" + ` uses the dev-database to re-run all migration files in the migration
```
Comment (a8m) [inline cmd/action/migrate.go:51]:
```suggestion
		Long: "'atlas migrate diff' uses the dev-database to re-run all migration files in the migration
```

File: cmd/action/migrate.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+	MigrateCmd.PersistentFlags().StringSliceVarP(&MigrateFlags.Schemas, migrateFlagSchema, "", nil, "set schema names")
+	MigrateCmd.PersistentFlags().BoolVarP(&MigrateFlags.Force, migrateFlagForce, "", false, "force a command to run on a broken migration directory state")
+	MigrateDiffCmd.Flags().StringVarP(&MigrateFlags.DevURL, migrateDiffFlagDevURL, "", "", "[driver://username:password@address/dbname?param=value] select a data source using the DSN format")
+	MigrateDiffCmd.Flags().StringVarP(&MigrateFlags.ToURL, migrateDiffFlagTo, "", "", "[driver://username:password@address/dbname?param=value] select a data source using the DSN format")
+	MigrateDiffCmd.Flags().SortFlags = false
+	cobra.CheckErr(MigrateDiffCmd.MarkFlagRequired(migrateDiffFlagDevURL))
+	cobra.CheckErr(MigrateDiffCmd.MarkFlagRequired(migrateDiffFlagTo))
+}
+
+func CmdMigrateDiffRun(cmd *cobra.Command, args []string) {
```
Comment (a8m) [inline cmd/action/migrate.go:74]: fix lint

File: cmd/action/migrate.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+		Args: cobra.MaximumNArgs(1),
+		Run:  CmdMigrateDiffRun,
+	}
+)
+
+func init() {
+	RootCmd.AddCommand(MigrateCmd)
+	MigrateCmd.AddCommand(MigrateDiffCmd)
+	MigrateCmd.PersistentFlags().StringVarP(&MigrateFlags.DirURL, migrateFlagDir, "", "file://migrations", "select migration directory using DSN format")
+	MigrateCmd.PersistentFlags().StringSliceVarP(&MigrateFlags.Schemas, migrateFlagSchema, "", nil, "set schema names")
```
Comment (a8m) [inline cmd/action/migrate.go:65]: Move it close to the `ToURL`.

File: cmd/action/migrate.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+	MigrateCmd.PersistentFlags().BoolVarP(&MigrateFlags.Force, migrateFlagForce, "", false, "force a command to run on a broken migration directory state")
+	MigrateDiffCmd.Flags().StringVarP(&MigrateFlags.DevURL, migrateDiffFlagDevURL, "", "", "[driver://username:password@address/dbname?param=value] select a data source using the DSN format")
+	MigrateDiffCmd.Flags().StringVarP(&MigrateFlags.ToURL, migrateDiffFlagTo, "", "", "[driver://username:password@address/dbname?param=value] select a data source using the DSN format")
+	MigrateDiffCmd.Flags().SortFlags = false
+	cobra.CheckErr(MigrateDiffCmd.MarkFlagRequired(migrateDiffFlagDevURL))
+	cobra.CheckErr(MigrateDiffCmd.MarkFlagRequired(migrateDiffFlagTo))
```
Comment (a8m) [inline cmd/action/migrate.go:71]: 💯

File: cmd/action/migrate.go [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+	dir, err := dir()
+	cobra.CheckErr(err)
+	// Get a state reader for the desired state.
+	desired, err := to(cmd.Context(), dev)
+	cobra.CheckErr(err)
+	// Only create one file per plan.
+	tfmt, err := migrate.NewTemplateFormatter(nameTmpl, contentTmpl)
+	cobra.CheckErr(err)
+	// Plan the changes and create a new migration file.
+	pl := migrate.NewPlanner(dev, dir, migrate.WithFormatter(tfmt))
```
Comment (a8m) [inline cmd/action/migrate.go:88]:
```suggestion
	tf, err := migrate.NewTemplateFormatter(nameTmpl, contentTmpl)
	cobra.CheckErr(err)
	// Plan the changes and create a new migration file.
	pl := migrate.NewPlanner(dev, dir, migrate.WithFormatter(tf))
```

File: cmd/action/migrate.go [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+	var n string
+	if len(args) > 0 {
+		n = args[0]
+	}
+	plan, err := pl.Plan(cmd.Context(), n, desired)
+	cobra.CheckErr(err)
+	return
```
Comment (a8m) [inline cmd/action/migrate.go:95]: rm

File: cmd/action/migrate.go [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+	// Write the plan to a new file.
+	cobra.CheckErr(pl.WritePlan(plan))
+}
+
+// dir returns a migrate.Dir to use as migration directory. For now only local directories are supported.
+func dir() (migrate.Dir, error) {
```
Comment (a8m) [inline cmd/action/migrate.go:101]: If the `dir` is generic, don't you think it's preferred to name it as `from`?

File: cmd/action/migrate.go [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+	parts := strings.SplitN(MigrateFlags.DirURL, "://", 2)
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid dsn %q", MigrateFlags.DirURL)
```
Comment (a8m) [inline cmd/action/migrate.go:104]:
```suggestion
		return nil, fmt.Errorf("invalid dir url %q", MigrateFlags.DirURL)
```

File: cmd/action/migrate.go [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+	}
+}
+
+// to returns a migrate.StateReader for the given to flag.
+func to(ctx context.Context, d *Driver) (migrate.StateReader, error) {
+	parts := strings.SplitN(MigrateFlags.ToURL, "://", 2)
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid dsn %q", MigrateFlags.ToURL)
+	}
+	schemas := ApplyFlags.Schema
```
Comment (a8m) [inline cmd/action/migrate.go:120]: MigrateFlags.

File: cmd/action/migrate.go [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+	default:
+		return nil, fmt.Errorf("unsupported driver %q", parts[0])
+	}
+}
+
+// to returns a migrate.StateReader for the given to flag.
+func to(ctx context.Context, d *Driver) (migrate.StateReader, error) {
+	parts := strings.SplitN(MigrateFlags.ToURL, "://", 2)
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid dsn %q", MigrateFlags.ToURL)
```
Comment (a8m) [inline cmd/action/migrate.go:118]:
```suggestion
		return nil, fmt.Errorf("invalid to url %q", MigrateFlags.ToURL)
```

File: cmd/action/migrate.go [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+	}
+	schemas := ApplyFlags.Schema
+	if n, err := SchemaNameFromURL(ctx, parts[0]); n != "" {
+		cobra.CheckErr(err)
+		schemas = append(schemas, n)
+	}
+	switch parts[0] {
+	case "file": // hcl file
```
Comment (a8m) [inline cmd/action/migrate.go:126]: I like the `file://` format, but maybe we should support also `-f` to simplify it? Same for `dir` and `from`. But note the 2 are mutually exclusive.

File: cmd/action/migrate.go [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+		f, err := ioutil.ReadFile(parts[1])
+		if err != nil {
+			return nil, err
+		}
+		realm := new(schema.Realm)
```
Comment (a8m) [inline cmd/action/migrate.go:131]:
```suggestion
		realm := &schema.Realm{}
```
Just to stay consistent, I try to not use `new` in the codebase, but it's just a suggestion to you.

File: cmd/action/migrate.go [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+				return s + ";"
+			}
+			return s
+		},
+	}
+	nameTmpl = template.Must(template.New("name").Funcs(funcMap).Parse(
+		"{{ now }}{{ with .Name }}_{{ . }}{{ end }}.sql",
+	))
+	contentTmpl = template.Must(template.New("content").Funcs(funcMap).Parse(
+		"{{ range .Changes }}{{ println (sem .Cmd) }}{{ end }}",
```
Comment (a8m) [inline cmd/action/migrate.go:177]: Add the Comment as well? If exists of course.

**Review by a8m — COMMENTED** [review]
Neat! 💯

File: cmd/action/migrate.go (hunk 6 above)
Comment (masseelch) [reply cmd/action/migrate.go:65]: `--schemas` is a global flag. `-to` is not :-(

File: cmd/action/migrate.go [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+	// Only create one file per plan.
+	tfmt, err := migrate.NewTemplateFormatter(nameTmpl, contentTmpl)
+	cobra.CheckErr(err)
+	// Plan the changes and create a new migration file.
+	pl := migrate.NewPlanner(dev, dir, migrate.WithFormatter(tfmt))
+	var n string
+	if len(args) > 0 {
+		n = args[0]
+	}
+	plan, err := pl.Plan(cmd.Context(), n, desired)
```
[the remark this answers is not in the document]
Comment (masseelch) [reply cmd/action/migrate.go:93]: Just the version is. If the user does not want to give a name it is fine.

File: cmd/action/migrate.go (hunk 10 above)
Comment (masseelch) [reply cmd/action/migrate.go:101]: We already have a `--from` flag in the `atlas schema diff` command and that is something else than here. Therefore I chose `--dir` to distinguish it from each other.

File: cmd/action/migrate.go (hunk 14 above)
Comment (masseelch) [reply cmd/action/migrate.go:126]: I guess we could fallback to `file://` if no source type is given, if the users request it in the future?

**Review by a8m — APPROVED** [review]

File: cmd/action/migrate.go [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+	pl := migrate.NewPlanner(dev, dir, migrate.WithFormatter(tf))
+	var name string
+	if len(args) > 0 {
+		name = args[0]
+	}
+	plan, err := pl.Plan(cmd.Context(), name, desired)
+	cobra.CheckErr(err)
+	// Write the plan to a new file.
+	cobra.CheckErr(pl.WritePlan(plan))
+	// TODO(masseelch): clean up dev after reading the state from migration dir.
```
Comment (a8m) [inline cmd/action/migrate.go:103]: Should be done in the “migrate” imo.

File: sql/migrate/migrate.go [hunk 19]
```diff
@@ -130,6 +130,13 @@ func Schema(s *schema.Schema) StateReader {
 	})
 }
 
+// Conn returns a StateReader for a Driver.
+func Conn(drv Driver, opts *schema.InspectRealmOption) StateReader {
```
Comment (a8m) [inline sql/migrate/migrate.go:134]: 👌🏻

File: sql/migrate/migrate.go [hunk 20]
```diff
@@ -219,12 +226,18 @@ func GlobStateReader(dir Dir, drv Driver, glob string) StateReaderFunc {
...
-			b, err := io.ReadAll(f)
-			f.Close()
-			if err != nil {
-				return nil, err
+			sc := bufio.NewScanner(f)
+			for sc.Scan() {
+				t := sc.Text()
+				if !strings.HasPrefix(strings.TrimSpace(t), "--") {
+					if _, err := drv.ExecContext(ctx, t); err != nil {
+						f.Close()
```
Comment (a8m) [inline sql/migrate/migrate.go:234]: Maybe defer?

**Review by a8m — APPROVED** [review]
😍

File: cmd/action/migrate.go (hunk 18 above)
Comment (masseelch) [reply cmd/action/migrate.go:103]: I think it fits more into the state-reader, but let's discuss tomorow.

**Pull request merged**